File: exercism/haskell/yacht/src/Yacht.hs
module Yacht (yacht, Category(..)) where
import Data.List (nub, sort)

data Category = Ones
              | Twos
              | Threes
              | Fours
              | Fives
              | Sixes
              | FullHouse
              | FourOfAKind
              | LittleStraight
              | BigStraight
              | Choice
              | Yacht

yacht :: Category -> [Int] -> Int
yacht Ones dice = count 1 dice
yacht Twos dice = (count 2 dice) * 2
yacht Threes dice = (count 3 dice) * 3
yacht Fours dice = (count 4 dice) * 4
yacht Fives dice = (count 5 dice) * 5
yacht Sixes dice = (count 6 dice) * 6
yacht FullHouse dice =
  case sorted of
    (2:3:[]) -> sum dice
    _ -> 0
  where counts = map (flip count dice) [1..6]
        sorted = sort (filter (>=1) counts)

yacht FourOfAKind dice
  | snd max >= 4 = 4 * fst max
  | otherwise = 0
  where counts = map (flip count dice) [1..6]
        max = mostDices counts

yacht LittleStraight dice
  | and [(length (nub dice) == 5), ((sum dice) == 15)] = 30
  | otherwise = 0

yacht BigStraight dice
  | and [(length (nub dice) == 5), ((sum dice) == 20)] = 30
  | otherwise = 0

yacht Choice dice = sum dice

yacht Yacht (x:xs)
  | all (== x) xs = 50
  | otherwise = 0

count :: Int -> [Int] -> Int
count el xs = foldr (\ e acc -> if e == el then acc+1 else acc) 0 xs

mostDices :: [Int] -> (Int, Int)
mostDices xs = mostDices' xs 0 0 0

mostDices' :: [Int] -> Int -> Int -> Int -> (Int, Int)
mostDices' [] _ y idx = (idx+1, y)
mostDices' (x:xs) i y idx
  | x > y = mostDices' xs (i + 1) x i
  | otherwise = mostDices' xs (i + 1) y idx

File: exercism/elixir/yacht/lib/yacht.ex
defmodule Yacht do
  @type category ::
          :ones
          | :twos
          | :threes
          | :fours
          | :fives
          | :sixes
          | :full_house
          | :four_of_a_kind
          | :little_straight
          | :big_straight
          | :choice
          | :yacht

  @doc """
  Calculate the score of 5 dice using the given category's scoring method.
  """
  @spec score(category :: category(), dice :: [integer]) :: integer
  def score(:ones, dice), do: filter_sum(dice, 1)
  def score(:twos, dice), do: filter_sum(dice, 2)
  def score(:threes, dice), do: filter_sum(dice, 3)
  def score(:fours, dice), do: filter_sum(dice, 4)
  def score(:fives, dice), do: filter_sum(dice, 5)
  def score(:sixes, dice), do: filter_sum(dice, 6)

  def score(:full_house, dice) do
    case Enum.sort(dice) do
      [x, x, x, y, y] when x != y -> Enum.sum(dice)
      [y, y, x, x, x] when x != y -> Enum.sum(dice)
      _ -> 0
    end
  end

  def score(:four_of_a_kind, dice) do
    case Enum.sort(dice) do
      [y, y, y, y, _] -> y * 4
      [_, x, x, x, x] -> x * 4
      _ -> 0
    end
  end

  def score(:little_straight, dice) do
    case Enum.sort(dice) do
      [1, 2, 3, 4, 5] -> 30
      _ -> 0
    end
  end

  def score(:big_straight, dice) do
    case Enum.sort(dice) do
      [2, 3, 4, 5, 6] -> 30
      _ -> 0
    end
  end

  def score(:choice, dice), do: Enum.sum(dice)

  def score(:yacht, [x, x, x, x, x]), do: 50
  def score(:yacht, _), do: 0

  defp filter_sum(enum, elem), do: Enum.filter(enum, &(&1 == elem)) |> Enum.sum()
end
